### PR TITLE
ensure system python path is not pushed to front of PYTHONPATH

### PR DESCRIPTION
**Pull request by grondo**
First attempt at solving #1999.

Ensure that a system python path is never pushed to the front of `PYTHONPATH` by installing links to Flux directories installed under `$pyexecdir` in a non-standard path. The `flux(1)` cmddriver can then prepend *this* path to `PYTHONPATH` instead of a possible system path which could result in overriding user preference. As discussed in #1999, this approach still allows `import flux` using the system python since the modules are still really installed in the default `pyexecdir`.

This seems to work as intended and doesn't break any tests (I think).  However, definitely one of our python gurus should check it out.

### CODE REVIEW

**Comment by grondo** [comment]
This branch breaks `make distcheck`. I'll look into it.

**Comment by grondo** [comment]
Ok, fixed the issue with `make distcheck` (wish I knew the "right" way to install a symlink with automake, the current approach does it by hand)

File: configure.ac [hunk 1]
```diff
@@ -356,6 +358,13 @@ AC_SUBST(fluxcmddir)
 AS_VAR_SET(fluxlibdir, $libdir/flux)
 AC_SUBST(fluxlibdir)
 
+# Target of PYTHONPATH set by flux(1) cmddriver, so flux(1)
+# doesn't inadvertently insert system python paths (or any
+# other pytho path for that matter) first in PYTHONPATH.
```
Comment (chu11) [inline configure.ac]: typo `pytho`

File: configure.ac [hunk 2]
```diff
@@ -251,10 +251,12 @@ if test "$ac_python_ldflags_set_by_user" != "true"; then
     PYTHON_LDFLAGS="-l$ac_python_library"
   fi
 fi
+AS_VAR_SET(fluxpyexecdir, $pyexecdir/flux)
+AC_SUBST(fluxpydir)
```
Comment (chu11) [inline configure.ac]: should this be `fluxpyexecdir`?

Comment (grondo) [reply configure.ac]: Good catch! Actually that is a rebase error and those lines need to be removed (`fluxpyexecdir` was from an earlier version of the PR)

File: configure.ac (hunk 1 above)
Comment (grondo) [reply configure.ac]: Thanks!

**Comment by grondo** [comment]
Thanks @chu11! Rebased addressing your comments.

**Comment by chu11** [comment]
LGTM, but I think a python expert should probably take a glance too (@trws or @SteVwonder)

**Comment by SteVwonder** [comment]
Thanks @grondo for putting this together!

Gave this a spin, and it works well.  My only concern is what happens when we configure flux for multiple pythons.

Right now, this is what happens when you configure/install for python3.7 and then again for python2.7:

```
# tree $PREFIX/lib

lib
├── flux
│   └── python
│       ├── flux -> $PREFIX/lib/python2.7/site-packages/flux
│       └── _flux -> $PREFIX/lib/python2.7/site-packages/_flux
└── python2.7
    └── site-packages
        ├── flux
        ├── _flux
        ├── <every other python2.7 system-installed package>
└── python3.7
    └── site-packages
        ├── flux
        ├── _flux
        ├── <every other python3.7 system-installed package>
```

So `flux python` will work, launching the python that flux was configured with and not pollute the PYTHONPATH with a system path, but if a user wants to use a python3.7 interpreter and `import flux`, they cannot do so easily without including a "polluted" system path in their PYTHONPATH (i.e., /usr/lib/flux/python3.7/site-packages).

Can we tweak `fluxpylinkdir` to have the major/minor version numbers at the end of the directory name so that this workaround works when there are multiple python versions?

**Comment by grondo** [comment]
> but if a user wants to use a python3.7 interpreter and import flux, they cannot do so easily without including a "polluted" system path in their PYTHONPATH (i.e., /usr/lib/flux/python3.7/site-packages).

Isn't this the same as any other python package installed on the system? How do users normally import a system installed package without "polluting" their PYTHONPATH?

I don't think I had the use case in mind that users would want to use `/usr/lib/flux/python` as a way to "isolate" the flux python bindings by manually adding that path to PYTHONPATH. This solution is only to isolate the flux PYTHONPATH for flux itself.

Since `flux(1)` can only be compiled to support a single version of python, I thought it appropriate to just have a single directory.

That being said, there probably isn't any harm in adding the python version to `/usr/lib/flux/python`, assuming the python version is easily available in `configure.ac` this should be trivial.

**Comment by SteVwonder** [comment]
> Isn't this the same as any other python package installed on the system? How do users normally import a system installed package without "polluting" their PYTHONPATH?

Yes, it is how other system-installed python packages are added.  Normally, if a user wants to import a specific package without pulling in system packages, they will install it locally with `pip install --user PACKAGE_NAME` (or install it with pip into a virtualenv).  Currently, we don't put our python bindings up on PyPI (the Python Package Index), so that's not an option for our users.   Adding our bindings to PyPI is probably the better long-term solution (created https://github.com/flux-framework/flux-core/issues/2145 to track that), but I'm wonder if this would be sufficient for the short-term.

> This solution is only to isolate the flux PYTHONPATH for flux itself.

Yeah, this PR solves #1999 perfectly.  So if we want to leave how it is, that is fine.

> assuming the python version is easily available in configure.ac this should be trivial.

It "should be" :tm: as simple as referencing `PYTHON_VERSION` [[1]](https://www.gnu.org/software/automake/manual/html_node/Python.html)

**Comment by grondo** [comment]
Yes, as you hinted, the following patch (pushed as a separate commit for now) just seems to work:
```diff
diff --git a/configure.ac b/configure.ac
index 57d45b781..224ff43c3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -360,7 +360,7 @@ AC_SUBST(fluxlibdir)
 # doesn't inadvertently insert system python paths (or any
 # other python path for that matter) first in PYTHONPATH.
 #
-AS_VAR_SET(fluxpylinkdir, $fluxlibdir/python)
+AS_VAR_SET(fluxpylinkdir, $fluxlibdir/python$PYTHON_VERSION)
 AC_SUBST(fluxpylinkdir)
 
 AS_VAR_SET(fluxmoddir, $libdir/flux/modules)
```

Let me know what you think and I can squash that one down.

**Comment by grondo** [comment]
Great I'll squash that last commit

**Comment by grondo** [comment]
squashed.

**Comment by SteVwonder** [comment]
LGTM!  Ready for merging?

**Comment by grondo** [comment]
> LGTM! Ready for merging? 

Yes, thanks!

**Comment by grondo** [comment]
Thanks!

**Pull request merged**